### PR TITLE
Captcha

### PR DESCRIPTION
**Pull request by git-jls**
When creating new tickets as a non signed in user, this will add a captcha above the submit button.
To use the captcha two keys are needed:

```
RECAPTCHA_PUBLIC_KEY
RECAPTCHA_PRIVATE_KEY
```

These keys are set by the recaptcha.rb file in the config/initializer directory and are read from the secrets.yml file.

This pull request will include two testing keys, but for production a set of keys need to be created at the [official website](https://www.google.com/recaptcha) and exported to the environment directly, like so:

```
export RECAPTCHA_PUBLIC_KEY="[YOUR_KEY]"
export RECAPTCHA_PRIVATE_KEY="[YOUR_KEY]"
```

This pull request also includes an entry about captcha in the README.md and two tests that check for breakage.

This should fix [#228](https://github.com/ivaldi/brimir/issues/228)

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -84,3 +84,6 @@ gem 'fancybox2-rails', '~> 0.2.8'
 
 # gravatar for user avatar images
 gem 'gravatar_image_tag'
+
+# Captcha for brimir
+gem "recaptcha", require: "recaptcha/rails"
```
Comment (frenkel) [inline Gemfile]: Doesn't it work without the `require`?

File: README.md [hunk 2]
```diff
@@ -44,6 +44,13 @@ If you want to use LDAP, configure config/ldap.yml accordingly, then change the
     bin/rails console production
     u = User.new({ email: 'your@email.address', password: 'somepassword', password_confirmation: 'somepassword' }); u.agent = true; u.save!
 
+IMPORTANT Captcha Notice:
```
Comment (frenkel) [inline README.md]: Could you change this to a head? (Using `-----`) and rename it to "Configuring Captcha's" or something like that.

File: app/controllers/tickets_controller.rb [hunk 3]
```diff
@@ -30,9 +30,9 @@ def show
     @agents = User.agents
 
     draft = @ticket.replies
-        .where('user_id IS NULL OR user_id = ?', current_user.id)
-        .where(draft: true)
-        .first
+      .where('user_id IS NULL OR user_id = ?', current_user.id)
+      .where(draft: true)
+      .first
```
Comment (frenkel) [inline app/controllers/tickets_controller.rb]: You're undoing correct indentation here.

File: app/controllers/tickets_controller.rb [hunk 4]
```diff
@@ -55,9 +55,9 @@ def show
       format.eml do
         begin
           send_file @ticket.raw_message.path(:original),
-              filename: "ticket-#{@ticket.id}.eml",
-              type: 'text/plain',
-              disposition: :attachment
+            filename: "ticket-#{@ticket.id}.eml",
+            type: 'text/plain',
+            disposition: :attachment
```
Comment (frenkel) [inline app/controllers/tickets_controller.rb]: You're undoing correct indentation here.

File: app/controllers/tickets_controller.rb [hunk 5]
```diff
@@ -84,7 +84,7 @@ def index
     respond_to do |format|
       format.html do
         @tickets = @tickets.paginate(page: params[:page],
-            per_page: current_user.per_page)
+                                     per_page: current_user.per_page)
```
Comment (frenkel) [inline app/controllers/tickets_controller.rb]: You're undoing correct indentation here.

File: app/controllers/tickets_controller.rb [hunk 6]
```diff
@@ -193,5 +190,30 @@ def create
 
       format.js { render }
     end
+
+  end
+
+  private
```
Comment (frenkel) [inline app/controllers/tickets_controller.rb]: `protected` is preferred, so a subclass can override or use these as well.

File: app/controllers/tickets_controller.rb [hunk 7]
```diff
@@ -159,26 +159,23 @@ def create
...
-            render 'create'
-          else
-            redirect_to ticket_url(@ticket), notice: I18n::translate(:ticket_added)
-          end
-
+        # if recpatcha is enabled, set conditionals
+        if Ticket.recaptcha_keys_present? && verify_recaptcha 
+          ticket_respond_to_html
+        elsif !Ticket.recaptcha_keys_present?
+          ticket_respond_to_html
```
Comment (frenkel) [inline app/controllers/tickets_controller.rb]: This seems simpler:
```
if !Ticket.recaptcha_keys_present?
  ticket_respond_to_html
elsif verify_recaptcha
  ticket_respond_to_html
end
```

File: app/controllers/tickets_controller.rb [hunk 8]
```diff
@@ -159,26 +159,23 @@ def create
       @ticket = Ticket.new(ticket_params)
     end
 
-    if !@ticket.nil? && @ticket.save
-      NotificationMailer.incoming_message(@ticket, params[:message])
+    if Ticket.recaptcha_keys_present? && verify_recaptcha
+      ticket_save_and_sent_notification
+    elsif !Ticket.recaptcha_keys_present?
+      ticket_save_and_sent_notification
```
Comment (frenkel) [inline app/controllers/tickets_controller.rb]: This seems simpler:
```
if !Ticket.recaptcha_keys_present?
  ticket_save_and_sent_notification
elsif verify_recaptcha
  ticket_save_and_sent_notification
end
```

File: app/models/ticket.rb [hunk 9]
```diff
@@ -164,6 +164,15 @@ def to
     to_email_address.try :email
   end
 
+  def self.recaptcha_keys_present?
+    if Recaptcha.configuration.public_key.blank? &&
```
Comment (frenkel) [inline app/models/ticket.rb]: `||` seems the correct one here, not `&&`

Comment (frenkel) [inline app/models/ticket.rb]: You could also write this method as:

```
def self.recaptcha_keys_present?
  !Recaptcha.configuration.public_key.blank? &&
      !Recaptcha.configuration.private_key.blank?
end
```

File: config/initializers/recaptcha.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  Recaptcha.configure do |config|
+    config.public_key   = Rails.application.secrets[:recaptcha_public_key]
+    config.private_key  = Rails.application.secrets[:recaptcha_private_key]
+  end
+else
+  Recaptcha.configure do |config|
+    config.public_key   = Rails.application.secrets[:recaptcha_public_key]
+    config.private_key  = Rails.application.secrets[:recaptcha_private_key]
+  end
+end
```
Comment (frenkel) [inline config/initializers/recaptcha.rb]: I don't see why the if/else is needed, both do exactly the same thing?

File: test/controllers/tickets_controller_test.rb [hunk 11]
```diff
@@ -76,9 +76,9 @@ class TicketsControllerTest < ActionController::TestCase
     assert_no_difference 'ActionMailer::Base.deliveries.size' do
       assert_no_difference 'Ticket.count' do
         post :create, ticket: {
-            from: 'invalid',
-            content: '',
-            subject: '',
+          from: 'invalid',
+          content: '',
+          subject: '',
```
Comment (frenkel) [inline test/controllers/tickets_controller_test.rb]: You're undoing correct indentation here.

File: test/controllers/tickets_controller_test.rb [hunk 12]
```diff
@@ -93,9 +93,9 @@ class TicketsControllerTest < ActionController::TestCase
     assert_difference 'ActionMailer::Base.deliveries.size', User.agents.count do
       assert_difference 'Ticket.count' do
         post :create, ticket: {
-            from: 'test@test.nl',
-            content: @ticket.content,
-            subject: @ticket.subject,
+          from: 'test@test.nl',
+          content: @ticket.content,
+          subject: @ticket.subject,
```
Comment (frenkel) [inline test/controllers/tickets_controller_test.rb]: You're undoing correct indentation here.

File: test/controllers/tickets_controller_test.rb [hunk 13]
```diff
@@ -111,9 +111,9 @@ class TicketsControllerTest < ActionController::TestCase
     assert_difference 'ActionMailer::Base.deliveries.size', User.agents.count do
       assert_difference 'Ticket.count', 1 do
         post :create, ticket: {
-            from: 'test@test.nl',
-            content: @ticket.content,
-            subject: @ticket.subject,
+          from: 'test@test.nl',
+          content: @ticket.content,
+          subject: @ticket.subject,
         }
```
Comment (frenkel) [inline test/controllers/tickets_controller_test.rb]: You're undoing correct indentation here.

File: test/controllers/tickets_controller_test.rb [hunk 14]
```diff
@@ -166,7 +166,7 @@ class TicketsControllerTest < ActionController::TestCase
 
     # should have selected same outgoing address as original received
     assert_select 'option[selected="selected"]' +
-        "[value=\"#{email_addresses(:brimir).id}\"]"
+      "[value=\"#{email_addresses(:brimir).id}\"]"
```
Comment (frenkel) [inline test/controllers/tickets_controller_test.rb]: You're undoing correct indentation here.

File: test/controllers/tickets_controller_test.rb [hunk 15]
```diff
@@ -320,6 +320,32 @@ class TicketsControllerTest < ActionController::TestCase
     end
   end
 
+  test 'shown captcha should not break new ticket page when not signed in' do
+    # Stub keys Recaptcha
+    Recaptcha.configuration.public_key = '6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI'
+    Recaptcha.configuration.private_key = '6LeIxAcTAAAAAGG-vFI1TnRWxMZNFuojJ4WifJWe'
```
Comment (frenkel) [inline test/controllers/tickets_controller_test.rb]: These are loaded using your initializer and secrets file, you can remove these two lines.

File: app/controllers/tickets_controller.rb (hunk 8 above)
Comment (alexhouse) [inline app/controllers/tickets_controller.rb]: Surely that can be simplified further?

```ruby
    if !Ticket.recaptcha_keys_present? || verify_recaptcha
        ticket_save_and_sent_notification
    end
```

File: config/initializers/recaptcha.rb (hunk 10 above)
Comment (frenkel) [inline config/initializers/recaptcha.rb]: Could you still fix this?

File: app/controllers/tickets_controller.rb [hunk 16]
```diff
@@ -159,26 +159,23 @@ def create
...
-          if current_user.nil?
-            render 'create'
-          else
-            redirect_to ticket_url(@ticket), notice: I18n::translate(:ticket_added)
-          end
-
+        # if recpatcha is enabled, set conditionals
+        if !Ticket.recaptcha_keys_present?
+          ticket_respond_to_html
+        elsif verify_recaptcha
```
Comment (frenkel) [inline app/controllers/tickets_controller.rb]: I didn't realize that you're requiring signed in users to enter the captcha as well. Could you hide the captcha and not verify it for them? Only not signed in users should enter the captcha as an anti-spam measure.

**Comment by frenkel** [comment]
Thanks! I'll fix the last issue myself (no recaptcha keys means you can't create tickets anymore)

**Pull request merged**